From 50aa1c8a0a0d32eee584c8b20338e59750d0904f Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 13 Mar 2021 19:39:16 +0100
Subject: [PATCH 097/179] genirq: irq_pipeline: inband_irq helpers must not
 trace the interrupt state

First of all, since the inband_irq helpers are folded into local_irq*
calls which already trace the inband interrupt state via the
trace_hardirqs* helpers, tracing such state in the inband_irq*
routines too is entirely redundant.

In addition, inband_irq* helpers are also used by raw_local_irq*
forms, which are supposed not to trace the interrupt state. Typically,
lockdep assumes this when verifying the sanity of the global interrupt
state vs lockdep interrupt state, i.e.

	  /* irqs are (virtually) on */
	  raw_local_irq_save(flags);
	  check_flags(flags);
	  ...

In this case, check_flags() is supposed to find out that @flags
represents an unstalled context, which should match hardirqs_enabled
== 1, because raw_local_irq_save() should have skipped the
trace_hardirqs* machinery.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/irq/pipeline.c | 9 +--------
 1 file changed, 1 insertion(+), 8 deletions(-)

diff --git a/kernel/irq/pipeline.c b/kernel/irq/pipeline.c
index 4e3003612..7addeab41 100644
--- a/kernel/irq/pipeline.c
+++ b/kernel/irq/pipeline.c
@@ -280,7 +280,6 @@ static void __inband_irq_enable(void)
 	flags = hard_local_irq_save();
 
 	unstall_inband_nocheck();
-	trace_hardirqs_on();
 
 	p = this_inband_staged();
 	if (unlikely(stage_irqs_pending(p) && !in_pipeline())) {
@@ -329,7 +328,6 @@ notrace void inband_irq_disable(void)
 {
 	check_inband_stage();
 	stall_inband_nocheck();
-	trace_hardirqs_off();
 }
 EXPORT_SYMBOL(inband_irq_disable);
 
@@ -358,13 +356,8 @@ EXPORT_SYMBOL(inband_irqs_disabled);
  */
 trace_on_debug unsigned long inband_irq_save(void)
 {
-	unsigned long flags;
-
 	check_inband_stage();
-	flags = test_and_stall_inband_nocheck();
-	trace_hardirqs_off();
-
-	return flags;
+	return test_and_stall_inband_nocheck();
 }
 EXPORT_SYMBOL(inband_irq_save);
 
-- 
2.38.1

